cpp/src/arrow/type.cc

```diff
@@ -327,14 +709,27 @@ Result<std::shared_ptr<Field>> Field::MergeWith(const Field& other,
     return Copy();
   }
 
-  if (options.promote_nullability) {
-    if (type()->Equals(other.type())) {
-      return Copy()->WithNullable(nullable() || other.nullable());
+  auto maybe_promoted_type = MergeTypes(type_, other.type(), options);
+  if (!maybe_promoted_type.ok()) {
+    return maybe_promoted_type.status().WithMessage(
+        "Unable to merge: Field ", name(),
+        " has incompatible types: ", type()->ToString(), " vs ", other.type()->ToString(),
+        ": ", maybe_promoted_type.status().message());
+  }
+  auto promoted_type = *std::move(maybe_promoted_type);
+  if (promoted_type) {
+    bool nullable = nullable_;
+    if (options.promote_nullability) {
+      nullable = nullable || other.nullable() || type_->id() == Type::NA ||
+                 other.type()->id() == Type::NA;
+    } else if (nullable_ != other.nullable()) {
+      return Status::Invalid("Unable to merge: Field ", name(),
+                             " has incompatible nullability: ", nullable_, " vs ",
+                             other.nullable());
```

Comment:
We should probably change these errors to `TypeError`. It may slightly break user code, but for the better (so that one can distinguish between unsupported type casts and invalid value casts).